ISC26: update the basics section (#535)
spack/spack-tutorial: tutorial_basics.rst

Reviewer: I think that there's a simpler way of explaining this than with wildcards?

Author: I went with:
```
Pass a substring to narrow the list.
For example, to see all Python packages:
```

=========================================

This tutorial will provide a step-by-step guide for installing software with Spack.

A few fundamental ideas underpin everything that follows:

> 1. Spack builds software either from source or from prebuilt binaries.

Reviewer: ```suggestion
1. Spack can install software either from source or a prebuilt binary.
```

> 2. Spack keeps every configuration of a package isolated from every other, so many versions, compilers, and build options can coexist on the same machine.

Reviewer: ```suggestion
2. Spack isolates each package's installation. This allows many versions, compilers, and build options to coexist on the same machine.
```

> 3. Each configuration is identified by a hash of its full provenance.

Reviewer: ```suggestion
3. Each installation is identified by a hash of its full provenance.
```

4. Spack reuses an existing build whenever it can, instead of rebuilding from scratch.

Keep these points in mind, as we'll illustrate them with examples.

> We will begin by introducing the ``spack install`` command, highlighting the versatility of Spack's spec syntax and the flexibility it offers users.

Reviewer: ```suggestion
We will begin by introducing the ``spack install`` command, showcasing the flexibility of Spack's spec syntax.
```

   :language: console
   :lines: 1-6

Once we've found a package, installing it is as simple as typing ``spack install`` followed by its name:

.. code-block:: console

  $ spack install <package_name>

Let's install ``gmake``:

.. literalinclude:: outputs/basics/gmake.out
   :language: spec

In the output, the ``[e]`` marker denotes a package that was found on the system rather than built by Spack.
Spack's output lists ``gmake``, ``gcc``, ``gcc-runtime``, ``glibc``, and ``compiler-wrapper``:

> * ``gmake`` is the package we requested.

Reviewer: I like this being in bullet point format

All compilers Spack found are configured as *external packages*.
We'll cover externals in the "Spack Concepts" slides and in the :ref:`Configuration Tutorial <configs-tutorial>` later on.

**Spack can install software either from source or from a binary cache.**
We just built ``gmake`` from source.
To speed up the rest of the tutorial, let's add a binary cache:

.. literalinclude:: outputs/basics/mirror.out
   :language: console

From here on, the same ``spack install`` command will fetch a package from the cache when a matching binary exists and fall back to building from source when it doesn't.

---------------
The Spec Syntax
---------------

So far we've installed packages with their default configuration.
Spack's *spec syntax* is how we request a specific configuration of a package.
A *spec* describes a package together with any constraints we want to place on how it is built: its version, its build options, the compiler it uses, and even the configuration of its dependencies.
Each kind of constraint has its own sigil, which the subsections below introduce one at a time, building up from a bare package name to fully constrained dependency graphs.

^^^^^^^^
Versions
^^^^^^^^

> We can install multiple versions of the same package side by side.

Reviewer: ```suggestion
Spack can install multiple versions of the same package.
```
?

Before installing a specific version, let's check which versions of ``zlib-ng`` are available using the ``spack versions`` command.

.. literalinclude:: outputs/basics/versions-zlib.out
   :language: spec

> We select one with the ``@`` sigil:

Reviewer: ```suggestion
We specify versions with the ``@`` sigil:
```

   :language: spec

This installation is located separately from the previous one.
As described in the overview, this separation is fundamental to how Spack supports multiple configurations and versions of software packages simultaneously.

**The spec syntax is recursive**: any syntax we can specify for the "root" package we can also use for a dependency.
For example, since a compiler is just another dependency, we can pin its version with ``@``, just as we did for ``zlib-ng``:

.. literalinclude:: outputs/basics/zlib-gcc-14.out
   :language: spec

^^^^^^^^^^^^^^^^^^^^^^^
Transitive Dependencies
^^^^^^^^^^^^^^^^^^^^^^^

The ``^`` sigil can constrain any dependency of a root spec, whether direct or transitive.

We need a package with dependencies to try it on.
The ``tcl`` package depends on ``zlib-ng``, so let's preview how Spack would build ``tcl`` with constraints on its ``zlib-ng`` dependency using ``spack spec``; the ``-l`` flag adds each node's hash:

.. literalinclude:: outputs/basics/spec-tcl-zlib-clang.out
   :language: spec

This is the *concretized* spec: Spack has filled in every dependency, along with its version, variants, and compiler.

> In the output, ``[+]`` marks specs already installed, ``[e]`` those provided externally by the system, and ``[b]`` those available in a cache but not yet installed.

Reviewer: ```suggestion
In the output, ``[+]`` marks specs already installed, ``[e]`` marks those provided externally by the system, and ``[b]`` marks those available in a cache but not yet installed.
```

   :language: spec

As we saw when referring to builds by hash, every installed package has a distinct hash, so configurations that look alike in the default output still occupy separate installations.

``spack find`` can also show what each installed package depends on with the ``-d`` flag.
For example, here is the ``tcl`` we installed, shown with its dependency tree:

.. literalinclude:: outputs/basics/find-d-tcl.out
   :language: spec

The ``spack find`` command can also accept what we call "anonymous specs": expressions in spec syntax that do not contain a package name.
For example, ``spack find ^mpich`` will return every installed package that depends on MPICH.

.. literalinclude:: outputs/basics/find-dep-mpich.out
   :language: spec

The ``find`` command can show which packages were installed explicitly (rather than pulled in as a dependency) using the lowercase ``-x`` flag; the uppercase ``-X`` flag shows implicit installs only.
It can also show the path to which a package was installed using the ``-p`` flag.

.. literalinclude:: outputs/basics/find-px.out
   :language: spec

.. _basics-tutorial-trilinos:

-------------------
A Realistic Example
-------------------

> Now that we know the spec syntax and how to query installations, let's put them to work on a realistic package.

Reviewer: Maybe knit-picking, but I don't think I like the term "realistic" here, all of the other examples were realistic installations someone could do as well

Author: Let me know if this seems better: 
https://github.com/spack/spack-tutorial/pull/535/commits/8f70bdf31f085f8c30be8b68d2d81dba2634eb0c

Reviewer 2: @kshea21 is on vacation today. Edits look good to me.

.. literalinclude:: outputs/basics/uninstall-ambiguous.out
   :language: spec

As the error suggests, we can disambiguate with a more specific spec, refer to the exact build by its hash, or pass ``--all`` (or ``-a``) to remove every match.
Here we remove one of the two Trilinos builds by its hash:

.. literalinclude:: outputs/basics/uninstall-specific.out
   :language: spec

---------------------
Customizing Compilers
---------------------

In the :ref:`Installing Packages <basics-tutorial-install-packages>` section, we saw that Spack detects the compilers already on your ``PATH`` and configures them as external packages.
Spack can also install a compiler itself and then use it to compile other packages.

.. literalinclude:: outputs/basics/install-gcc-16.out
   :language: spec

Once installed, it appears with a ``[+]`` in the list of available compilers:

.. literalinclude:: outputs/basics/compilers-2.out
   :language: console

The ``gcc@16`` compiler is immediately available to use:

.. literalinclude:: outputs/basics/spec-zziplib.out
   :language: spec
   :lines: 1-2

> We won't need this compiler in the next chapter, so we'll uninstall it for now.

Reviewer: section
